src/cargo/ops/tree/format/mod.rs

```diff
@@ -0,0 +1,108 @@
+use self::parse::{Parser, RawChunk};
+use super::{Graph, Node};
+use anyhow::{anyhow, Error};
+use std::fmt;
+
+mod parse;
+
+enum Chunk {
+    Raw(String),
+    Package,
+    License,
+    Repository,
+    Features,
+}
+
+pub struct Pattern(Vec<Chunk>);
+
+impl Pattern {
+    pub fn new(format: &str) -> Result<Pattern, Error> {
+        let mut chunks = vec![];
+
+        for raw in Parser::new(format) {
+            let chunk = match raw {
+                RawChunk::Text(text) => Chunk::Raw(text.to_owned()),
+                RawChunk::Argument("p") => Chunk::Package,
+                RawChunk::Argument("l") => Chunk::License,
+                RawChunk::Argument("r") => Chunk::Repository,
+                RawChunk::Argument("f") => Chunk::Features,
+                RawChunk::Argument(a) => {
+                    return Err(anyhow!("unsupported pattern `{}`", a));
+                }
+                RawChunk::Error(err) => return Err(anyhow!("{}", err)),
+            };
+            chunks.push(chunk);
+        }
+
+        Ok(Pattern(chunks))
+    }
+
+    pub fn display<'a>(&'a self, graph: &'a Graph<'a>, node_index: usize) -> Display<'a> {
+        Display {
+            pattern: self,
+            graph,
+            node_index,
+        }
+    }
+}
+
+pub struct Display<'a> {
+    pattern: &'a Pattern,
+    graph: &'a Graph<'a>,
+    node_index: usize,
+}
+
+impl<'a> fmt::Display for Display<'a> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let node = self.graph.node(self.node_index);
+        match node {
+            Node::Package {
+                package_id,
+                features,
+                ..
+            } => {
+                let package = self.graph.package_for_id(*package_id);
+                for chunk in &self.pattern.0 {
+                    match *chunk {
+                        Chunk::Raw(ref s) => fmt.write_str(s)?,
```

Comment:
Nowadays this can probably be modernized with modern match syntax that avoids `ref` and such